internal/vcs/git/archive.go

```
@@ -21,27 +20,35 @@ const (
 
 // ArchiveReader streams back the file contents of an archived git repo.
 func ArchiveReader(
-	ctx context.Context,
-	repoName api.RepoName,
-	options gitserver.ArchiveOptions,
-) (io.ReadCloser, error) {
-	return gitserver.DefaultClient.Archive(ctx, repoName, options)
-}
-
-func ArchiveReaderWithSubRepo(
 	ctx context.Context,
 	checker authz.SubRepoPermissionChecker,
-	repo *types.Repo,
+	repo api.RepoName,
 	options gitserver.ArchiveOptions,
 ) (io.ReadCloser, error) {
+	options, err := validateOptions(options)
+	if err != nil {
+		return nil, err
+	}
 	if authz.SubRepoEnabled(checker) {
-		enabled, err := authz.SubRepoEnabledForRepoID(ctx, checker, repo.ID)
-		if err != nil {
+		if enabled, err := authz.SubRepoEnabledForRepo(ctx, checker, repo); err != nil {
 			return nil, errors.Wrap(err, "sub-repo permissions check:")
+		} else if enabled {
+			filteredFiles, err := LsFiles(ctx, checker, repo, api.CommitID(options.Treeish), options.Paths...)
+			if err != nil {
+				return nil, errors.Wrap(err, "LsFiles in ArchiveReader")
+			}
+			options.Paths = filteredFiles
```

this list can be too large. I imagine sub-repo permissions are common on large monorepos, which means this will normally fail.

tar is a streaming format so you can filter out. This is more difficult for zip, but doable since in practice the format generated will look like a streaming file.

In theory you could batch teh calls, but that is also tricky for combining. Another approach is to implement a perm aware git archive on gitserver (go-git?). Another approach is always asking for tar which is streamable, but then creating a zip file here.

Taking a step back, where do we expose archives these days? My historical knowledge here is for extensions (which I think is no longer the case?) and src-cli's batch stuff? Maybe there is another approach here for those and we can hide the archive endpoint / fail it for sub-repo perms?

---

I like both of these ideas. I feel like requesting a tar from gitserver and then filtering/creating a zip file here makes the most sense. At the moment we aren't doing sub-repo permissions filtering inside gitserver, but rather in the `git` package that wraps calls to the gitserver api. The approach of requesting the tar and filtering here would be more in line with that approach, which is why I'm leaning in that direction.

---

I might for now (i.e. this PR) just fail early if the list of files is too long and then do a follow-up to implement the filtering in that case just for ease of review.

---

I looked into go-git and it doesn't look like archiving is currently supported: https://sourcegraph.com/github.com/go-git/go-git/-/blob/COMPATIBILITY.md?L65  :/ 

Our original approach for sub-repo was to just fail anytime the archive endpoint is called when sub-repo perms are enabled. That might still be the best approach? I'm going back and forth on whether this is even worth it at this point. The reason I feel like it might be worth it is there are other areas of the codebase where we're attempting to call `git archive` with a long list of files (I think just symbols), so we're already needing to figure out how to handle this issue.

---

Yeah internally I believe everywhere we call archive it isn't on behalf of a user, but to analyse the repo to create an index / stats / etc. Those services then need to respect sub repo permissions themselves (And already do?). For example symbols does this as you point out, but will do it as "root". It will then filter out the symbols a user can't see for a user request.

I'm not actually sure where archives are used externally, the only memory I have is src-cli using them for batch changes. cc @eseliger who can comment on if archive is still needed. If not, I imagine just failing if the request is for a user makes a lot of sense to me since it is much simpler than actually making this work.

> I looked into go-git and it doesn't look like archiving is currently supported

Yeah you would have to implement this yourself, which would be relatively straightforward using its tree walker stuff. But to be honest, it would be easier to just implement the idea of always using tar and synthesizing a zip if requested.